[#235] 댓글: 실제 데이터 바인딩

jwonyLee opened the pull request:
## 개요

https://user-images.githubusercontent.com/15073405/142963584-e7d6fd8e-505f-43a0-9c92-036d27643fae.mp4




댓글 실제 데이터 바인딩

## 작업 사항

- 설정에 있던 댓글 화면 제거
- 댓글 저장 기능 구현
  - 이걸 구현하는 중간에 CommentEntity.createDate가 defaultDate로 저장되고 있는 걸 발견해서 현재 날짜로 저장하게 수정했습니다.
- 댓글 수정 기능 구현
  - 기존에 PostRepository에 있는 update 메서드는 Post 객체로 수정하기 때문에 PostEntity를 이용해 수정하는 메서드를 추가했습니다.
- 댓글 삭제 기능 구현
  - 댓글을 삭제 하기 전에 사용자에게 경고 알림을 띄웁니다.
- 실제 데이터 바인딩
- 게시물 화면으로 다시 돌아갔을 때 테이블뷰 업데이트
  - 댓글이 추가되거나, 삭제되고 나서 이전 화면으로 돌아갔을 때 `댓글 n개 모두 보기`가 갱신되어야 해서 추가했습니다.
- 기존에 댓글화면을 띄우기 위해서 PostEntity를 파라미터로 받고 있던 부분을, CommentViewModel로 받게 수정했습니다. CommentViewModel은 생성 시에 PostEntity가 필요합니다.

## 예외 사항

- 이전 화면으로 돌아갈 때 startIndexPath로 돌아가는 현상은 풀 리퀘스트 #243 에 있기 때문에 해결하지 않았습니다.

### Linked Issue

close #235

jwonyLee committed:
[#235] feat: 테스트 댓글 화면 제거, CommentVC로 이동

- 댓글 화면의 로직을 담당할 ViewModel 생성, Coordinator 리팩토링

jwonyLee committed:
[#235] feat: 본문, 댓글 바인딩

jwonyLee committed:
[#235] feat: 댓글 저장 기능 구현
- 저장 기능 구현 중에 CommentEntity.createDate가 defaultDate로 저장되고 있어서 현재 날짜로 저장되게 수정

jwonyLee committed:
[#235] feat: 댓글 삭제 기능 구현
- 댓글 삭제 전 사용자에게 경고 알림을 띄움

jwonyLee committed:
[#235] feat: 댓글 수정 기능 구현

jwonyLee committed:
[#235] feat: 게시물 화면으로 되돌아왔을 때 테이블뷰 새로고침

jwonyLee committed:
[#235] feat: 누락된 댓글 날짜 표시

gustn3965 commented on ThingLog/ViewController/Post/PostViewController.swift:
```diff
@@ -41,6 +41,13 @@ final class PostViewController: BaseViewController {
+            } else {
+                tableView.reloadData()
+            }
+        }
```
지금 이 코드가 맞나용..? 저는 댓글화면에서 게시물화면으로 돌아오면 아래 코드가 호출되지 않아서요.

gustn3965 commented on ThingLog/ViewModel/CommentViewModel.swift:
```diff
@@ -0,0 +1,80 @@
+    var contents: String? { postEntity.contents }
+    private(set) var postEntity: PostEntity
+    private(set) var repository: PostRepository = PostRepository(fetchedResultsControllerDelegate: nil)
+    private var comments: [CommentEntity] {
+        guard let commentEntities: [CommentEntity] = postEntity.comments?.allObjects as? [CommentEntity] else {
+            return []
+        }
+        let sortedComment: [CommentEntity] = commentEntities.sorted(by: {
+            $0.createDate ?? Date() < $1.createDate ?? Date()
+        })
+        return sortedComment
+    }
```
코멘트가 사실 엄청 많지는 않겠지만, 지금 이 프로퍼티가 연산프로퍼티로, `get**`메소드들이 호출될때 마다 위의 로직을 계속 거치게됩니다. 셀이 10개면 10번 위의로직이 계속 실행되죠. 스크롤해도 계속 실행됩니다.

그것보다는 `postEntity`가 업데이트 또는 삭제, 추가 등이 발생하고 난 후 `postEntity`를 위와 같이 정렬시킨 `postEntity`로 만들어주는 메소드로 따로 만들고,  이를 실행시켜 `comments` 프로퍼티에 할당하고,
`comments`는 `lazy var` 저장 프로퍼티로, 해당 메소드를 return값으로 받도록 하는건 어떨까요? 그럼 위의 로직은 업데이트만 발생할 떄만 실행되므로 비용이 훨씬 줄어든다고 생각합니다.

gustn3965 reviewed (commented):
고생하셨습니다. 코맨트들 확인해주세요 ~ 


추가적으로로, GUI에서는 콘텐츠내용이랑 댓글화면이 왼쪽으로 정렬되어있는데 현재 화면은 그렇지 않습니당!
![image](https://user-images.githubusercontent.com/48749182/142980643-e9e6045a-79d1-4705-a8e9-a399ff28e9e6.png)

jwonyLee committed:
[#235] ref: 피드백 반영
- 댓글 화면에서 게시물 화면으로 돌아왔을 때 데이터 reload 하는 부분 수정
- 실제 댓글 GUI 화면에 맞게 수정
- 가격 세자리 단위로 끊어서 표시
- sortedComments() 구현, 게시물 생성, 삭제 시 호출

jwonyLee commented:
피드백 주신 내용을 토대로 반영했습니다. 추가적으로 가격 표시 부분이 세자리 단위로 끊어서 표시하는 부분이 누락되어서 해당 부분도 추가했습니다.
- 댓글 화면에서 게시물 화면으로 돌아왔을 때 데이터 reload 하는 부분 수정
- 실제 댓글 GUI 화면에 맞게 수정
  - 게시물 화면에서 본문과 댓글 n개 모두 보기 간격이 맞지 않는 부분 수정
- 가격 세자리 단위로 끊어서 표시 → `Int64.toStringWithComma()`
- `sortedComments()` 구현, 게시물 생성, 삭제 시 호출

gustn3965 reviewed (approved):
LGTM 👍

jwonyLee committed:
Merge branch 'develop' into feature/235-comment
